sdks/python/apache_beam/runners/portability/artifact_service.py

```diff
@@ -0,0 +1,122 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Implementation of the Arrtifact{Staging,Retrieval}Service.
+
+The staging service here can be backed by any beam filesystem.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import hashlib
+import os
+import random
+import re
+
+from apache_beam.io import filesystems
+from apache_beam.portability.api import beam_artifact_api_pb2
+from apache_beam.portability.api import beam_artifact_api_pb2_grpc
+
+
+class BeamFilesystemArtifactService(
+    beam_artifact_api_pb2_grpc.ArtifactStagingServiceServicer,
+    beam_artifact_api_pb2_grpc.ArtifactRetrievalServiceServicer):
+
+  _DEFAULT_CHUNK_SIZE = 10 << 20  # 10mb
+
+  def __init__(self, root, chunk_size=_DEFAULT_CHUNK_SIZE):
+    self._root = root
+    self._chunk_size = chunk_size
+
+  def _sha256(self, string):
+    return hashlib.sha256(string.encode('utf-8')).hexdigest()
+
+  def _mkdir(self, retrieval_token):
+    dir = os.path.join(self._root, retrieval_token)
```

Comment:
This will not give the correct path separator, say on Windows, and using the GCS filesystem (this is the case for os.path.join throughout).

Reply:
Good catch. Fixed.